policy-controller/src/profiling.rs

```diff
@@ -0,0 +1,144 @@
+
+                line
+            })
+            .collect();
+
+        for line in lines {
+            writeln!(writer, "{}", line)?;
+        }
+
+        Ok(())
+    }
+
+}
+
+pub fn init() {
+    let guard = Arc::new(
+        ProfilerGuardBuilder::default()
+            .frequency(1000)
+            .blocklist(&["libc", "libgcc", "pthread", "vdso"])
+            .build()
+            .unwrap(),
+    );
+
+    let opt_query = warp::query::<QueryParams>()
+        .map(Some)
+        .or_else(|_| async { Ok::<(Option<QueryParams>,), std::convert::Infallible>((None,)) });
+
+    let pprof_report_endpoint =
+        warp::path("pprof")
+            .and(opt_query)
+            .map(move |params: Option<QueryParams>| {
+                let report = guard.report().build().unwrap(); 
+
+                match params {
+                    Some(obj) => {
+                        if obj.format == "flamegraph" {
+                            let mut file = Vec::new();
+                            report.flamegraph(&mut file).unwrap();
+                            Response::builder()
+                                .header("content-type", "image/svg+xml")
+                                .body(file)
+                                .unwrap()
+                        } else if obj.format == "proto" {
+                            let mut file = Vec::new();
+                            let profile = report.pprof().unwrap();
+                            profile.write_to_vec(&mut file).unwrap();
+                            Response::builder()
+                                .header("content-type", "application/octet-stream")
+                                .header("content-disposition", "attachment; filename=profile.pb")
+                                .body(file)
+                                .unwrap()
+                        } else if obj.format == "folded" {
+                            let my_report = MyReport(report);
+                            let mut file = Vec::new();
+                            my_report.folded(&mut file).unwrap();
+                            Response::builder()
+                                .header("content-type", "text/plain")
+                                .header("content-disposition", "attachment; filename=folded")
+                                .body(file)
+                                .unwrap()
+                        } else if obj.format == "tracing" {
+                            let file = Vec::new();
+                            let fmt_layer = fmt::Layer::default();
+                            let writer = Cursor::new(file.clone());
+                            let flame_layer = FlameLayer::new(writer);
+                            tracing_subscriber::registry()
+                                                .with(fmt_layer)
+                                                .with(flame_layer)
+                                                .init();
```

Comment:
... and this should be stored in a variable (e.g. `subscriber`) that then you need to set as the default global subscriber via `tracing::subscriber::set_global_default(subscriber).expect("Could not set global default");`

Again, this should go in the initialization phase. And this Warp endpoint should only care about delivering the contents of the buffer (`file`).